plugin/trino-redshift/src/main/java/io/trino/plugin/redshift/RedshiftClient.java

```diff
@@ -616,7 +632,10 @@ public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connect
             case Types.BIT: // Redshift uses this for booleans
                 return Optional.of(booleanColumnMapping());
 
-            // case Types.TINYINT: -- Redshift doesn't support tinyint
+            case Types.TINYINT:
+                // -- Redshift doesn't support tinyint, but requires
+                // tinyint data type mapping for CAST pushdown
+                return Optional.of(tinyintColumnMapping());
```

Comment:
If it doesn't support tinyint - Do we need to perform this coercion ? What if operators on tinyint doesn't work on Redshift ?

Reply:
For `getQueryRunner().execute("SELECT CAST(c_int AS TINYINT) FROM " + leftTable());`

Without this change: `toColumnMapping` maps `tinyint` to `VARCHAR` which fails at https://github.com/krvikash/trino/blob/92d956338dfa3d357483d8297fe06a4b2e6d94b5/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcRecordCursor.java#L91-L94


`
Caused by: io.trino.spi.TrinoException: Type mismatch: column handle has type tinyint but JdbcTypeHandle[jdbcType=-6, jdbcTypeName=Optional[tinyint], columnSize=Optional.empty, decimalDigits=Optional.empty, arrayDimensions=Optional.empty, caseSensitivity=Optional.empty] is mapped to varchar
`

With this change: `toColumnMapping` maps `tinyint` to `TINYINT`.

Trino's `TINYINT` type mapped to Redshift's `SMALLINT` type.

`EXPLAIN SELECT CAST(c_int AS TINYINT) FROM left_table_adsqw21pot`

```
MaterializedResult{rows=[[Trino version: testversion
Fragment 0 [SOURCE]
    Output layout: [pfgnrtd]
    Output partitioning: SINGLE []
    Output[columnNames = [_col0]]
    │   Layout: [pfgnrtd:tinyint]
    │   Estimates: {rows: ? (?), cpu: 0, memory: 0B, network: 0B}
    │   _col0 := pfgnrtd
    └─ TableScan[table = redshift:Query[SELECT CAST("c_int" AS smallint) AS "_pfgnrtd_0" FROM "testdb"."test_schema"."left_table_adsqw21pot"] columns=[_pfgnrtd_0:tinyint:tinyint]]
           Layout: [pfgnrtd:tinyint]
           Estimates: {rows: ? (?), cpu: ?, memory: 0B, network: 0B}
           pfgnrtd := _pfgnrtd_0:tinyint:tinyint

]], types=[varchar(606)], setSessionProperties={}, resetSessionProperties=[]}
```

Reply:
Just exploring alternative, what if we generate smallint type handle in the `toJdbcTypeHandle`?

Reply:
Yes, I tried that as well and It fails. Because the column type which gets used is the projection type after rewrite projection expression.

https://github.com/trinodb/trino/blob/c8f1bd00760316fee00837125ada36a80e3ac406/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java#L465

Reply:
But tinyint and smallint could have a different min and max value right ? How would we ensure that it works for value which exceeds tinyint range but a valid smallint range ? Since Redshift doesn't have a tinyint type - This coercion could be skipped right ?

Reply:
We map that column as `smallint` and read them as `tinyint` so things would fail at the time of reading ? What if the expression is hidden within the query like some sort of a filter and this column doesn't need to be read by the underlying system.

Reply:
Since TinyInt is not supported by Redshift - what if we could skip the pushdown for such operation

Reply:
> Since TinyInt is not supported by Redshift - what if we could skip the pushdown for such operation

Yes, we are removing tinyint pushdown